### ISSUE TITLE
stgit not finding user information

### ISSUE THREAD

**Issue by immae** [stacked-git/stgit#438]
Hello,

I’m running stgit version 2.4.5 . Since a recent change in the way I’m handling my git config, stgit became unable to get my user.name and user.email information:
```
repo $ stg new                                                                                                                                                            
error: author identity unknown; please configure `user.name` and `user.email`.
repo $ git config --get user.email                                                                                                                                        
immae@example.com
repo $ git config --get user.name
Immae
```

The change is as follows: my global git config used to have a `[user]` section.
Now I’m making use of a git feature to include conditional configuration:
```
[includeIf "hasconfig:remote.*.url:perso@gitlab.com:*/**"]
        path = ~/.config/git/perso.gitconfig
[includeIf "hasconfig:remote.*.url:work@gitlab.com:*/**"]
        path = ~/.config/git/work.gitconfig
```
and the work.gitconfig  / perso.gitconfig paths now contain the `[user]` section instead of the global git config.

This works just fine with normal git commands, but not with stgit. I suppose that stgit is doing a separate parsing of the git config file and misses that information. Is there any way to get that working?

Thanks in advance!

**Comment by jpgrayson** [comment]
StGit relies on gitoxide's gix-config library for parsing/interpreting git configuration. I successfully use `[includeIf "gitdir:..."]` in one of my environments, but have not used `[includeIf "hasconfig:..."]`. I suspect `gix-config` simply does not support `hasconfig` predicates yet.

A quick read of https://github.com/Byron/gitoxide/blob/main/gix-config/src/file/includes/types.rs#L122 seems to confirm that `gitdir` predicates are supported, but `hasconfig` is nowhere to be found.

@Byron: would love your thoughts on this.

**Comment by Byron** [comment]
Indeed, [`hasconfig` isn't yet supported](https://github.com/Byron/gitoxide/blob/3fd668454acece9d218dc7b5200836dce938efbc/crate-status.md?plain=1#L717), and it's [quite a special way](https://git-scm.com/docs/git-config#Documentation/git-config.txt-codehasconfigremoteurlcode) of including files.

After thinking about it, I believe that it would perfectly lend itself to being resolved as part of [this call](https://github.com/Byron/gitoxide/blob/f8ce3d0721b6a53713a9392f2451874f520bc44c/gix/src/config/cache/init.rs#L120), which already happens after most of the git configuration has been injected. Thus, it would work without any 'lookahead'. Of course, the devil is in the details here, but I am sure those can be figured out.

Since reading the git configuration faithfully is important to the correct functioning of any tool, I am prioritizing the implementation of `hasconfig` and will do so next time I get a chance. Thanks again for reeling me in - a thread like this is just what I need as motivation to implement it :).

PS: I have put it into [this PR](https://github.com/Byron/gitoxide/pull/1339) which is my catch-all that is merged from time to time, but also there to remind me what the big-picture goal is.

**Comment by immae** [comment]
Thanks for your answers! Indeed "gitdir:..." syntax works and I can live with that (although not as concisely as hasconfig would do) until hasconfig gets supported.

I’d love to help but I’m not fluent enough with rust yet to do anything about it
